src/Tools/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/SolutionState.cs

```diff
@@ -0,0 +1,259 @@
+            DiagnosticDescriptor defaultDiagnostic,
+            ImmutableArray<DiagnosticDescriptor> supportedDiagnostics,
+            ImmutableArray<string> fixableDiagnostics,
+            string defaultPath)
+        {
+            if (AllowMarkup == MarkupMode.None)
+            {
+                return (explicitDiagnostics.Select(diagnostic => diagnostic.WithDefaultPath(defaultPath)).ToOrderedArray(), sources.ToArray());
+            }
+
+            var sourceFiles = new List<(string filename, SourceText content)>();
+            var diagnostics = new List<DiagnosticResult>(explicitDiagnostics.Select(diagnostic => diagnostic.WithDefaultPath(defaultPath)));
+            foreach ((var filename, var content) in sources)
+            {
+                TestFileMarkupParser.GetSpans(content.ToString(), out var output, out IDictionary<string, IList<TextSpan>> namedSpans);
+                sourceFiles.Add((filename, content.Replace(new TextSpan(0, content.Length), output)));
+                if (namedSpans.Count == 0)
+                {
+                    // No markup notation in this input
+                    continue;
+                }
+
+                if (AllowMarkup == MarkupMode.Ignore)
+                {
+                    // The source contained markup, which was removed and ignored
+                    continue;
+                }
+
+                var sourceText = SourceText.From(output, content.Encoding, content.ChecksumAlgorithm);
+                foreach ((var name, var spans) in namedSpans)
+                {
+                    foreach (var span in spans)
+                    {
+                        var diagnostic = CreateDiagnosticForSpan(defaultDiagnostic, supportedDiagnostics, fixableDiagnostics, name, filename, sourceText, span);
+                        if (!diagnostic.HasValue)
+                        {
+                            continue;
+                        }
+
+                        diagnostics.Add(diagnostic.Value);
+                    }
+                }
+            }
+
+            return (diagnostics.ToOrderedArray(), sourceFiles.ToArray());
+        }
+
+        private DiagnosticResult? CreateDiagnosticForSpan(
+            DiagnosticDescriptor defaultDiagnostic,
+            ImmutableArray<DiagnosticDescriptor> supportedDiagnostics,
+            ImmutableArray<string> fixableDiagnostics,
+            string diagnosticId,
+            string filename,
+            SourceText content,
+            TextSpan span)
+        {
+            var linePositionSpan = content.Lines.GetLinePositionSpan(span);
+
+            DiagnosticResult diagnosticResult;
+            if (string.IsNullOrEmpty(diagnosticId))
+            {
+                if (defaultDiagnostic is null)
+                {
+                    throw new InvalidOperationException("Markup syntax can only omit the diagnostic ID if the first analyzer only supports a single diagnostic");
```

Comment:
📝 This restriction may be relaxed in the future, but for now I preserved the documented behavior.